### ISSUE TITLE
design: DIDComm Python API

### ISSUE THREAD

**Issue by ashcherbakov** [sicpa-dlab/didcomm-gemini#18]
## What
Define API and interfaces for Python DIDComm library matching https://hackmd.io/gIk7LWd_QZuzApNfZ4KEIA and be useable by ACAPy.

## Why
We need to define and review the API and interfaces and make sure that it's clear, simple, convenient and idiomatic before work on the implementation.

## Success Criteria
- Create a project for Python DIDComm
- Define required interfaces and API calls. It should not have any implementation, only stubs.
- Decide how to do private key management: pass it directly to the API calls, or create some abstractions such as wallets or secure enclaves. It makes sense to consider multiple options how private key management can be done and used by AcaPy and discuss this with the AcaPy users.
- Add API docs to the calls
- If there are any other diagrams/docs created while working on the task, make it available as a Wiki page in this repo.
- Review the API with the team
- Take into account the following when designing and reviewing the API:
   - API is clear, simple and convenient
   - API matches the latest version of the DID Comm spec https://identity.foundation/didcomm-messaging/spec/
   - API matches https://hackmd.io/gIk7LWd_QZuzApNfZ4KEIA
   - API is idiomatic for the language
   - API should be relatively easy to call from AcaPy

## Other Notes
Regarding private key and AcaPy:
- AcaPy currently uses Indy's Wallet for crypto operations (including pack/unpack/authcrypt/anoncrypt/etc.).
- Indy Wallet API doesn't expose private key and serves as a secure enclave there. Private keys are only used and accessible in internal modules
- So, we need to consider multiple options and decide how to deal with this. If we go with direct private key pass to the didcomm API, how can AcaPy be adapted to this. If we go with wallet interface for didcomm libs, how AcaPy can register Indy (or any other wallet) there.

**Comment by dhh1128** [comment]
Regarding the question about Indy's wallet and crypto operations: please see [my comments in issue #5](https://github.com/sicpa-dlab/didcomm-gemini/issues/5#issuecomment-879783969)

**Comment by ashcherbakov** [comment]
@dhh1128 @vimmerru 
Before we start working on a precise language-specific API, we need to define the basic principles of the DID Comm API and functionality. 
Please find a doc with possible options (language agnostic, just base ideas): https://hackmd.io/@n3TZrfxITW-lDRQPEJFjlA/SJkJgSNRu
I believe the options choice depends on
- whether we expect DID DOC parsing be part of DIDComm libs, or a separate library (either existing one or a new one but separate)
- whether DID Resolving (and DID DOC parsing) must be part of DIDComm libs from the DID Comm concept point of view. In other words, do we understand by DIDComm a JOSE-based message building and corresponding crypto, or DID resolving/parsing is also a conceptual and inseparable part of DID Comm.

If we think that DID DOC parsing is not inseparable part of DID Comm (which is actually true, as it follows DID-core spec, not DID Comm spec), then probably we can go with Option 2 (explicit API) or Option 4 (implicit and explicit API).

If we think that DID part is essential and inseparable part of DID Comm, then I would go with Option 4 (implicit and explicit API).

**Comment by dhh1128** [comment]
I want your option 2. I do not want DID Doc parsing to be part of DIDComm libraries. These libraries should be simple. DID Doc parsing is potentially complex, and I want these libraries to focus on the "Comm" part of "DIDComm".

**However**, I believe there are certain parts of DIDComm operations where we must resolve a DID before we can proceed. This means that although one of our DIDComm libraries shouldn't know how to do DID resolution and DID doc parsing itself, it depends on *something* to do that, right in the middle of certain DIDComm workflows. An example of this is that when I am decrypting a message from a DID, I want to know that the key that was used to encrypt it actually appeared in the `keyAgreement` section of the sender's DID document. If it didn't, DIDComm libs should throw an error, even if the crypto works.

In order to accomplish this, I believe the DIDComm libs must define an interface for DID resolution and DID doc parsing, into which implementations can be plugged in by whatever code is consuming the DIDComm libs. In other words, the DIDComm libraries know how to do the "Comm" part of DIDComm, AND they know how to call a library that provides the "DID" part -- but they don't know what specific library they will be calling for the DID functionality -- only how it will behave. It is then the job of the application code that uses the DIDComm libraries to decide what DID functionality it will plug in.

The way I imagine this to work is that, whenever DIDComm libraries encounter a situation where they need to know something from the DID impl, they call a method with a signature like this: `resolve(did, version_id) -> DIDDoc_interface_instance`. Then they can ask the DIDDoc_interface_instance questions like, "what's the DIDComm endpoint?" or "what are all the keyAgreement keys?" -- and the DIDDoc_interface_instance will provide an answer without revealing the internal data structures directly.

I believe the peer DID libs that we are planning to write for Gemini should be one possible pluggable implementation of the DID support. We should also write one that calls the Universal Resolver over a web service. Any other implementations of the DID features is out of scope.

**Comment by dhh1128** [comment]
Inviting @TelegramSam to comment...

**Comment by dhh1128** [comment]
Cross-referencing my comment about explicit APIs and key-versus-did arguments [here](https://github.com/sicpa-dlab/didcomm-gemini/issues/5#issuecomment-884162933).

**Comment by ashcherbakov** [comment]
Current design proposal we have a consensus about: https://hackmd.io/aD4TzKRhT1ma2ERl6EKU_g

**Comment by ashcherbakov** [comment]
PR is raised (WIP status, but some early feedback is welcome): https://github.com/sicpa-dlab/didcomm-python/pull/1

**Comment by ashcherbakov** [comment]
https://github.com/sicpa-dlab/didcomm-python/pull/1
- DID Comm API is defined (async version)
- DID and secrets resolving interfaces are defined
- demo example of API usage is added
- docstrings are added

Some details (such as exceptions raised by methods, etc.) can be changed during implementation.

Demo API scripts:
- [main demo](https://github.com/sicpa-dlab/didcomm-python/blob/1e1ced75ec24479cbe13949e563e9423eb9e769a/tests/demo/test_demo.py)
- [mediator demo](https://github.com/sicpa-dlab/didcomm-python/blob/1e1ced75ec24479cbe13949e563e9423eb9e769a/tests/demo/test_demo_forward.py)

**Comment by ashcherbakov** [comment]
@dbluhm @CluEleSsUK 
We did a number of modifications/improvements to the API, but I believe we have more or less final version now.
Can you please look at the proposed API, and either approve (so that we continue with implementation) or put your comments?

A [README](https://github.com/ashcherbakov/didcomm-python/blob/api-definition/README.md) may also be useful to undersand the high-level prnciples.
